### ISSUE TITLE
[Dashboard] Ray job endpoints become slow, when there are excessive loggers

### ISSUE THREAD

**Issue by jrosti** [ray-project/ray#39535]
### What happened + What you expected to happen

Excessive logging and job-tail functionality makes all job endpoints very slow. There are no known ways to recover, that is, killing all active client programs and stopping dashboard clients does not resume normal behaviour. 

If I submit following Python program few times:
```
# ray submit --entrypoint-resource "cpu8"  -- python logtoomuch.py
# cancel with ctrl-c
```
where logtoomuch.py is 
```
for i in range(10000000):
    print(f"foo {i}")
```
Ray job API becomes unresponsive, and I observe very long latencies:

```
'DELETE /api/jobs/raysubmit_neTM5YENdK2VGaZg HTTP/1.1' 200 174 bytes 135423865 us
/api/jobs/ HTTP/1.1' 200 2811 bytes 119013670 us '-' 'python-requests/2.31.0'
```
and the system does not fully recover even if I delete all jobs from the job table:
```
$ time ray job list
Job submission server address: http://10.216.160.195:8265/
[]

real	0m28.132s
```

Node is autoscaling, not sure if worker nodes on autoscaling cluster has impact to this. 
Notable, the event happens even if the supervisor nodes are scaled away.

### Versions / Dependencies

Ray versions 2.4.0 - 2.6.2.

On KubeRay. 

Various head node sizes, up to 8CPU / 27GB. 

### Reproduction script

```
# ray submit --entrypoint-resource  -- python logtoomuch.py
# cancel with ctrl-c
```
where logtoomuch.py is 
```
for i in range(10000000):
    print(f"foo {i}")
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by GeneDer** [comment]
@jrosti Just want to add some notes. Thanks for sharing detailed script and steps. I tried to reproduce this by running local Ray cluster and it doesn't have issue for nightly, 2.7.1, nor 2.4.0. Then I tried this using Kuberay cluster and I do see some issue once control + c is hit after the job is submitted, kubectl port-forward command will show the following error 
```
E1031 19:13:35.896505 1644960 portforward.go:391] error copying from local connection to remote stream: read tcp4 127.0.0.1:8265->127.0.0.1:39572: read: connection reset by peer
E1031 19:13:35.918482 1644960 portforward.go:378] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:8265->127.0.0.1:39572: write tcp4 127.0.0.1:8265->127.0.0.1:39572: write: broken pipe
```
and the entire dashboard becomes unreachable.
 
Can you confirm this is exactly the behavior you saw? And by restarting `kubectl port-forward`, the issue will go away?

**Comment by jrosti** [comment]
No, I doublechecked the behaviour by execing to the head and running commands over there. Additionally, dashboard access log shows execution time on the server side: '/api/jobs/ HTTP/1.1' 200 2811 bytes **119013670 us** '-' 'python-requests/2.31.0'

I have tried to move our workloads to setting, where driver/job supervisor is not on a node which is scaled away, and I haven't observed the job-table freeze since.

I guess that the trigger is:
* driver on a node which gets scaled away quickly after job exits
* some combination of logging and triggering the log tailing function on the dashboard same time when scaledown happens

I can try to add some debugging logs to the dashboard to see what it actually waits.

**Comment by jrosti** [comment]
Hi, 

I am able to reproduce issue with 2.7.0 with a condition that I define an entrypoint resource to a node which is scaled away first by Ray autoscaler, an later the underlying node by GKE. The delay does not occur before the GKE node scaling occurs.

I produced timing trace from `list_jobs` in job_head.py which calls `get_all_jobs` in `common.py`

All time is consumed in:
```
        driver_jobs, submission_job_drivers = await get_driver_jobs(
            self._dashboard_head.gcs_aio_client
        ) 
```
which takes 111 seconds. 

Everything else what happens in the function takes 1 second.

**Comment by jrosti** [comment]
Which turns out to wait in `get_all_job_info` in `job/utils.py`, and it goes to GCS Server:

gcs_server.out
```
[2023-11-06 22:17:09,134 I 9 9] (gcs_server) gcs_job_manager.cc:149: Getting all job info.
[2023-11-06 22:18:02,627 W 9 9] (gcs_server) gcs_job_manager.cc:227: Failed to get is_running_tasks from core worker: GrpcUnavailable: RPC Error message: failed to connect to all addresses; last error: UNKNOWN: ipv4:10.216.162.67:10003: Failed to connect to remote host: Connection timed out; RPC Error details: 
[2023-11-06 22:18:02,627 W 9 9] (gcs_server) gcs_job_manager.cc:227: Failed to get is_running_tasks from core worker: GrpcUnavailable: RPC Error message: failed to connect to all addresses; last error: UNKNOWN: ipv4:10.216.162.67:10003: Failed to connect to remote host: Connection timed out; RPC Error details: 
[2023-11-06 22:18:14,558 W 9 9] (gcs_server) gcs_job_manager.cc:227: Failed to get is_running_tasks from core worker: GrpcUnavailable: RPC Error message: failed to connect to all addresses; last error: UNKNOWN: ipv4:10.216.162.131:10003: tcp handshaker shutdown; RPC Error details: 
[2023-11-06 22:18:14,558 I 9 9] (gcs_server) gcs_job_manager.cc:185: Finished getting all job info.
```

**Comment by jrosti** [comment]
And, as I suspected, before the large delays started to occur

```
[2023-11-06 21:57:23,745 W 9 9] (gcs_server) gcs_job_manager.cc:227: Failed to get is_running_tasks from core worker: GrpcUnavailable: RPC Error message: failed to connect to all addresses; last error: UNKNOWN: ipv4:10.xxxx:10003: Failed to connect to remote host: Connection refused; RPC Error details: 
```

I see connection refused messages, which are instantaneously observed as a failure.

**Comment by jrosti** [comment]
Since Ray has observed that all nodes are dead, it should have marked all jobs as dead, so I hope that `void GcsJobManager::HandleGetAllJobInfo` in `gcs_job_manager.cc` goes into this branch:

```
      if (data.second.is_dead()) {
        reply->mutable_job_info_list(i)->set_is_running_tasks(false);
        core_worker_clients_.Disconnect(worker_id);
        (*num_processed_jobs)++;
        ;
        try_send_reply();
      } else {
```

^ but here I'm curious that why it calls `.Disconnect`? Does that communicate with dead worker? 

However, not able to proceed further since I haven't set compilation environment for the Ray GCS Server and can't inject it to my cluster.

**Comment by jrosti** [comment]
For me it looks clear that GCS client tries to communicate with dead worker, and in case the networking is so that it goes to TCP timeouts instead of refused, it will cause large delays. 

Hope this helps!

**Comment by GeneDer** [comment]
Thanks for the info @jrosti! I haven't go through all the code path you mentioned. But maybe it's just a matter of setting a timeout. Will look into it a bit more!

**Comment by jrosti** [comment]
Found this commit https://github.com/ray-project/ray/commit/5ac06e9b5daee18d206726fcdd5c89be633c60d1 which was landed to 2.8.0.

After upgrading, I was not able to reproduce the job table freeze anymore. Additionally, the fix in above commit seems to match perfectly to observations above.

**Comment by jrosti** [comment]
Fixed in 2.8.0.

**Comment by GeneDer** [comment]
Awesome! Thanks @jrosti Feel free to reopen the issue if you have other question!
